study-group/Toast-Implementation-Studies.md

```diff
@@ -70,25 +70,190 @@ The `<paper-toast>` holds onto information about how it was dismissed the last t
+      </div>
+    );
+  }
+}
+```
+
+`react-toastify` uses the explicit component pattern, by having the developer render a React `ToastContainer` component (the recommendation is to put this in the application root).
+This single component will render all toasts called from JavaScript and can be configured with global options.
+However, the library allows the developer a lot of freedom, supporting multiple `ToastContainer`s as well as no `ToastContainer`, though one is recommended.
+
+When the `toast()` is called with a message and a props object, it is displayed in the toast container, with the toast props overriding the `ToastContainer` props, like so:
+
+```js
+class App extends Component {
+  // Creating a toast that will close in 10000ms.
+  tenSecondToast = () => toast("This toast will self destruct in 10...", { autoClose: 10000 });
+  // Creating a toast that will close in 5000ms (the default).
+  normalToast = () => toast("This is a normal toast");
+
+  render(){
+    return (
+      <div>
+        <button onClick={this.tenSecondToast}>Close after 10000ms seconds</button>
+        <button onClick={this.normalToast}>Close after 5000ms (default)</button>
+        <ToastContainer autoClose={5000} />
+      </div>
+    );
+  }
+}
+```
+
+For some of the common controls, the toast can be configured like so:
+
+```js
+class App extends Component {
+  topRightWarning = () => toast('Warning!', {
+    type: toast.TYPE.WARNING,
+    position: toast.POSITION.TOP_RIGHT,
+    onClose: ({ props }) => window.alert('I warned you') // props: all props passed to rendered toast component
+  });
+
+  render() {
+    return (
+      <button onClick={this.topRightWarning}>Warn me</button>
+      <ToastContainer autoClose={5000} />
+    );
+  }
+}
+```
+
+`react-toastify` supports built-in themes for toasts, such as `success`, `info`, and `warning`, among others.
+It also provides intuitive positioning, that can be set from the toast or configured on the `ToastContainer`.
+Additionally, callbacks like `onClose` and `onOpen` allow the toast to have intelligent effects when the user interacts with it.
+
+Another extensible behavior `react-toastify` supports is updating a toast:
+
+```js
+class Update extends Component {
+  toastId = null;
+
+  notify = () => this.toastId = toast("Hello", { autoClose: false });
+
+  update = () => toast.update(this.toastId, { type: toast.TYPE.INFO, autoClose: 5000 });
+
+  render(){
+      return (
+        <div>
+          <button onClick={this.notify}>Notify</button>
+          <button onClick={this.update}>Update</button>
+        </div>
+      )
+  }
+}
+```
+
+As in this example, the `toastId` gives the developer control over the toast beyond configuration and display.
+
+### Notable Features & Details
+
+#### Toast ID
+the `toast()` function returns an ID of the toast it creates, which can be used to act upon that toast.
+For instance, with the toast ID the toast can be used with `dismiss()`, `update()`, and `isActive()`.
```

Comment:
I wonder what the motivation is, for using IDs instead of JavaScript objects with methods.

Reply:
I'm not sure, I'm going to add a note to the takeaways to make sure that's noted